backend/pubsub.js

```diff
@@ -1,127 +1,295 @@
+    ;['close', 'error', 'message', 'ping', 'pong'].forEach((eventName) => {
+      socket.on(eventName, (...args) => {
+        console.debug(`[pubsub] Event '${eventName}' on socket ${socket.id}`, ...args)
+        const customHandler = socket.server.customSocketEventHandlers[eventName]
+        const defaultHandler = (defaultSocketEventHandlers: Object)[eventName]
 
-      // https://github.com/swissmanu/primus-responder
-      spark.on('request', function (req, done) {
         try {
-          const { type, data: { contractID } } = req
-          const success = reply(SUCCESS, { type, id })
-          console.log(bold(`[pubsub] REQUEST '${type}' from '${id}'`), req)
-          switch (type) {
-            case SUB:
-              if (spark.rooms().indexOf(contractID) === -1) {
-                spark.join(contractID, function () {
-                  spark.room(contractID).except(id).write(req)
-                  done(success)
-                })
-              } else {
-                console.log(`[pubsub] ${id} already subscribed to: ${contractID}`)
-                done(success)
-              }
-              break
-            case UNSUB:
-              spark.room(contractID).except(id).write(req)
-              spark.leave(contractID, () => done(success))
-              break
-            case PUB:
-              spark.room(contractID).except(id).write(req)
-              break
-            default:
-              console.error(bold.red(`[pubsub] client ${id} didn't give us a valid type!`), req)
-              spark.leaveAll()
-              done(reply(ERROR, `invalid type: ${type}`))
-              spark.end()
+          if (defaultHandler) {
+            defaultHandler.call(socket, ...args)
           }
-        } catch (err) {
-          logger(err)
-          done(reply(ERROR, err))
+          if (customHandler) {
+            customHandler.call(socket, ...args)
+          }
+        } catch (error) {
+          logger(error)
         }
       })
-      spark.on('leaveallrooms', (rooms) => {
-        console.log(bold.yellow(`[pubsub] ${id} leaveallrooms`))
-        // this gets called on spark.leaveAll and 'disconnection'
-        rooms.forEach(contractID => {
-          primus.room(contractID).write(reply(UNSUB, { contractID, id }))
-        })
-      })
-
-      spark.on('data', function (data) {
-        console.log(bold.red('[pubsub] received UNHANDLED DATA from client:', data))
-      })
     })
+  },
+  error (error: Error) {
+    console.log('[pubsub] Server event: error', error)
+  },
+  headers () {
+  },
+  listening () {
+    console.log('[pubsub] Server event: listening')
+  }
+}
 
-    primus.on('disconnection', function (spark) {
-      // the spark that disconnected
-      console.log(bold.yellow(`[pubsub] ${spark.id} disconnection`))
-    })
-  })
+// Default handlers for server-side client socket events.
+// The `this` binding refers to the connected `ws` socket object.
```

Comment:
It's the web socket object form the `ws` library, not the native WebSocket which only exists in the frontend (and in Deno).
But yes a link to `ws` apidocs can be added somewhere.

Reply:
Although we are also currently using many other dependencies without including links to their documentation, for example `chalk` in this module.

Reply:
> But yes a link to ws apidocs can be added somewhere.

Yeah, see, you made it clear that I was confused what this was referring to, so I bet other developers might get confused too. A link would be super helpful then.

> Although we are also currently using many other dependencies without including links to their documentation, for example chalk in this module.

True... it's just that when `chalk` is used, there's a clear variable that is bound from an import, and you can trace it directly, whereas with these handlers it gets a bit more confusing because it's not clear where the `this` is bound just by reading the code. It's not easy to trace, so hence the documentation links on each one of them would be greatly helpful.